### PR TITLE
Make emojis larger when multiple are sent at the same time

### PR DESCRIPTION
**Pull request by eVoloshchak**
### Details
Previously emojis in messages were larger than text only if you send message consisting of only one emoji. This PR ensures that if message consists only of emojis, they are displayed larger that text, regardless of the length.

### Fixed Issues

https://github.com/Expensify/App/issues/8867

### Tests

1. Send a message that contains only emojis
2. Verify that emojis in this message are larger that the text (27pt)
3. Send a message that contains both text and emojis
4. Verify that emojis and text are small and have the same font size

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [ ] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [x] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<summary><h4>PR Reviewer Checklist</h4></summary>

- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

### QA Steps

1. Send a message that contains only emojis
2. Verify that emojis in this message are larger that the text (27pt)
3. Send a message that contains both text and emojis
4. Verify that emojis and text are small and have the same font size
#### Web

Screenshot from 2022-05-06 15-42-29
#### Mobile Web

Screenshot_20220506-154436

#### Desktop

I'm unable to launch the desktop app due to [this](https://expensify.slack.com/archives/C01GTK53T8Q/p1650494818552169?thread_ts=1650471398.505469&cid=C01GTK53T8Q) bug. It is not present if you downgrade to certain node and npm versions, but since I use MacinCloud, I'm unable to do so. So if someone with a mac could test it and upload a video, I would highly appreciate it.

#### iOS

Simulator Screen Shot - iPod touch (7th generation) - 2022-05-06 at 18 57 20

#### Android

Screenshot_20220506-154621

### CODE REVIEW

File: src/libs/EmojiUtils.js [hunk 1]
```diff
@@ -70,6 +70,30 @@ function isSingleEmoji(message) {
...
+ * Validates that this message contains only emojis
+ *
+ * @param {String} message
+ * @returns {Boolean}
+ */
+function containsOnlyEmojis(message) {
+    const match = message.match(CONST.REGEX.EMOJIS);
+    if (match) {
+        const codes = [];
+
```
Comment (Julesssss) [inline src/libs/EmojiUtils.js:83]: Could you please match the isSingleEmoji function by returning early? 

Also, I wonder if the two functions could be combined in any way?

Comment (eVoloshchak) [reply src/libs/EmojiUtils.js]: > Could you please match the isSingleEmoji function by returning early?

Done 
> Also, I wonder if the two functions could be combined in any way?

You mean `isSingleEmoji` and `containsOnlyEmojis`?
The only part the share is `const match = message.match(CONST.REGEX.EMOJIS)`, we could move it into a separate method

File: src/styles/styles.js [hunk 2]
```diff
@@ -1042,7 +1042,7 @@ const styles = {
         textDecorationLine: 'none',
     },
 
-    singleEmojiText: {
+    multipleEmojisText: {
```
Comment (Santhosh-Sellavel) [inline src/styles/styles.js:1045]: Could rename it as`onlyEmojisText` or `emojisText`  as the style applies for only text with emojis, it doesn't matter single or multiple?

**Review by Santhosh-Sellavel — CHANGES_REQUESTED** [review]

File: src/styles/styles.js (hunk 2 above)
Comment (eVoloshchak) [reply src/styles/styles.js]: > Could rename it as`onlyEmojisText` or `emojisText` as the style applies for only text with emojis, it doesn't matter single or multiple?

Updated

File: src/libs/EmojiUtils.js (hunk 1 above)
Comment (Julesssss) [reply src/libs/EmojiUtils.js]: Thanks. Okay, forget the second part then.

**Comment by Santhosh-Sellavel** [comment]
Even if there are spaces in between Emojis should they be small?

@eVoloshchak  @Julesssss

Screenshot 2022-05-10 at 9 57 09 PM

**Comment by Santhosh-Sellavel** [comment]
#### PR Reviewer Checklist

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

**Comment by Santhosh-Sellavel** [comment]
@eVoloshchak 
> I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)

Add failure steps i.e which is adding text along with emoji will make them smaller

**Comment by Julesssss** [comment]
> Even if there are spaces in between Emojis should they be small?

Hmm, nice edgecase! I missed that.

Part of me thinks that yes a space character could be an exception -- but it's a bit of a rabbit hole as I'm sure we could think of other exceptions too. Ultimately I don't think it matters too much either way

**Review by Julesssss — DISMISSED** [review]
Tested well for me on all platforms.

**Comment by Julesssss** [comment]
CC @tgolen who may want to review, as you have essentially already reviewed

File: src/styles/styles.js [hunk 3]
```diff
@@ -1042,7 +1042,7 @@ const styles = {
         textDecorationLine: 'none',
     },
 
-    singleEmojiText: {
+    onlyEmojisText: {
         fontSize: variables.fontSizeSingleEmoji,
         lineHeight: variables.fontSizeSingleEmojiHeight,
```
Comment (Santhosh-Sellavel) [inline src/styles/styles.js:1047]: NAB! these variables could be renamed too.

**Review by Santhosh-Sellavel — APPROVED** [review]
I left a few edge cases here https://github.com/Expensify/App/pull/8896#issuecomment-1122629076.

Otherwise Looks good & test's well for me!
cc: @Julesssss

**Review by Santhosh-Sellavel — CHANGES_REQUESTED** [review]
@eVoloshchak  Can you handle these cases thanks! https://github.com/Expensify/App/pull/8896#issuecomment-1122640256

File: src/libs/EmojiUtils.js [hunk 4]
```diff
@@ -70,6 +70,31 @@ function isSingleEmoji(message) {
...
+    const codes = [];
+    _.map(match, emoji => _.map(getEmojiUnicode(emoji).split(' '), (code) => {
+        if (code !== CONST.EMOJI_INVISIBLE_CODEPOINT) {
+            codes.push(code);
+        }
+        return code;
+    }));
+
+    const messageCodes = _.filter(_.map([...message.replace(/ /g, '')], char => getEmojiUnicode(char)), string => string.length > 0 && string !== CONST.EMOJI_INVISIBLE_CODEPOINT);
```
Comment (tgolen) [inline src/libs/EmojiUtils.js:94]: Don't forget to please add a comment for why the spread operator is necessary.

**Review by tgolen — CHANGES_REQUESTED** [review]

**Comment by eVoloshchak** [comment]
Updated

File: src/pages/home/report/ReportActionItemFragment.js [hunk 5]
```diff
@@ -97,6 +97,13 @@ const ReportActionItemFragment = (props) => {
                 );
             }
 
+            const differByLineBreaksOnly = props.fragment.html.replaceAll('<br />', ' ') === props.fragment.text;
```
Comment (tgolen) [inline src/pages/home/report/ReportActionItemFragment.js:100]: Can you please add a comment to the code which explains why this logic is necessary?

**Review by tgolen — CHANGES_REQUESTED** [review]
Looking good! Just this one last thing.

File: src/pages/home/report/ReportActionItemFragment.js (hunk 5 above)
Comment (eVoloshchak) [reply src/pages/home/report/ReportActionItemFragment.js:103]: > Can you please add a comment to the code which explains why this logic is necessary?

Updated

**Review by tgolen — DISMISSED** [review]
Thank you!

**Comment by tgolen** [comment]
all you @Santhosh-Sellavel

**Comment by Santhosh-Sellavel** [comment]
Yeah will check this today.

**Review by Julesssss — DISMISSED** [review]
Re-approving. Tests well 👍

File: src/libs/EmojiUtils.js [hunk 6]
```diff
@@ -70,6 +70,34 @@ function isSingleEmoji(message) {
...
+    _.map(match, emoji => _.map(getEmojiUnicode(emoji).split(' '), (code) => {
+        if (code !== CONST.EMOJI_INVISIBLE_CODEPOINT) {
+            codes.push(code);
+        }
+        return code;
+    }));
+
+    // Emojis are stored as multiple characters, so we're using spread operator
+    // to iterate over the actual emojis, not just characters that compose them
+    const messageCodes = _.filter(_.map([...message.replace(/ /g, '').replaceAll('\n', '')],
```
Comment (Santhosh-Sellavel) [inline src/libs/EmojiUtils.js:96]: This is not DRY, `message.replace(/ /g, '').replaceAll('\n', '')` computing this twice.

**Review by Santhosh-Sellavel — CHANGES_REQUESTED** [review]
Just a small change can we clean this up, thanks!

File: src/libs/EmojiUtils.js (hunk 6 above)
Comment (eVoloshchak) [reply src/libs/EmojiUtils.js]: > This is not DRY, `message.replace(/ /g, '').replaceAll('\n', '')` computing this twice.

All done

**Review by Santhosh-Sellavel — APPROVED** [review]
Looks good to me, tests well! 

cc: @tgolen @Julesssss

**Review by Julesssss — APPROVED** [review]
Thanks all, looks good to me 👍

**Comment by Santhosh-Sellavel** [comment]
I think we are good to merge this cc: @tgolen

**Comment by marcaaron** [comment]
Interesting FYI that these changes appear to be breaking mobile web for me on iOS simulator running iOS 13 - my device running iOS 15 is not affected. Pretty sure we are not tracking support for mobile web on all iOS versions so unsure if this counts as a "regression" - but gonna fix it.

**Comment by marcaaron** [comment]
`replaceAll()` seems to not be available in Safari iOS 13.

**Comment by eVoloshchak** [comment]
@marcaaron, that's a good catch, thank you!
`replaceAll()` is indeed [unavailable](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll#browser_compatibility) on iOS 13.3 and below, so we can just use `replace()` with `//g` regex instead. 
I submitted a PR that fixes this: https://github.com/Expensify/App/pull/9063

**Comment by Santhosh-Sellavel** [comment]
@marcaaron I think we should capture this to avoid this happening again.

**Comment by marcaaron** [comment]
Thanks, I'm working on a fix already so feel free to close your PR @eVoloshchak.

>@marcaaron I think we should capture this somewhere to avoid this happening again.

One solution I can think of is to...

1. Create a change in `expensify-common` `Str.replaceAll()` and use a polyfill there
2. Enforce this method is used with a lint rule that prohibits using `replaceAll()` and prefers using `Str.replaceAll()`

**Comment by Santhosh-Sellavel** [comment]
> One solution I can think of is to...
> 
> 1. Create a change in `expensify-common` `Str.replaceAll()` and use a polyfill there
> 2. Enforce this method is used with a lint rule that prohibits using `replaceAll()` and prefers using `Str.replaceAll()`

Sounds good to me!

**Pull request merged**